src/visitors/perf_visitor.hpp

```diff
@@ -199,94 +198,48 @@ class PerfVisitor: public AstVisitor {
 
     void visit_program(ast::Program& node) override;
 
-    void visit_plot_block(ast::PlotBlock& node) override {
-        measure_performance(&node);
-    }
+    void visit_plot_block(ast::PlotBlock& node) override;
 
     /// skip initial block under net_receive block
```

Comment:
general question : 
* I see that moving definitions to `.cpp` avoids include into headers.
* It perfectly make sense for common headers like `ast` or `ast_visitor`.
* But, if those headers are not used in other code but in their own compilation units, does it make sense to move single line functions to .cpp?
* I don't see any advantage in cases like `perf_visitor`. Separating declaration and definition for single liner function seems just code bloat.

Reply:
> * But, if those headers are not used in other code but in their own compilation units, 

I am not sure to follow. Isn't a header useless if only useful to its compilation unit? Isn't its purpose to expose symbols used by other compilation units?

Most visitors expose an API that deal with AST top nodes like Ast, Program, ...
From what I saw in this project, most code that uses visitors only manipulate these nodes, so there is not point overwhelm them with `ast.hpp`

> does it make sense to move single line functions to .cpp?

From a performance point of view, inline virtual member methods are not inlined by the compiler. And this is the case in most of the time here.

> I don't see any advantage in cases like perf_visitor. Separating declaration and definition for single liner function seems just code bloat.

Code bloating is subjective, and I tend to not agree with you. IMO having `inline` method definitions inside a class definition is a necessary evil to have more efficient code. I wish headers were clean, and user oriented. For instance, I would even mark the `visit_` method `private` because completely out of context for the final class user.

But I respect the beliefs of the code maintainers :) Let me know if you prefer me to rollback the changes in `perf_visitor`

Reply:
> I am not sure to follow. Isn't a header useless if only useful to its compilation unit? Isn't its purpose to expose symbols used by other compilation units?
> Code bloating is subjective, and I tend to not agree with you. I

Ah, I meant in the following context:

```
# this is foo.hpp

#include "bar.hpp"

class Foo {
    void some_simple_one_liner_method() {
           bar_obj->some_method();
    }

   // other complex methods that will go into foo.cpp
   void some_complex_method();
}
```
In this example, `bar` forward declaration is not sufficient. So to avoid include `bar.hpp`, we move `some_simple_one_liner_method` to `foo.cpp`. Given that:
* foo.hpp is not included by other files 
* do we need to move `one liner` methods to `.cpp` ?

 thought that is duplication. While writing this, I see an example where `foo.hpp` will be included in `main.cpp` where `Foo` will be instantiated and it will pull in `bar.hpp`! I see the point.

> For instance, I would even mark the visit_ method private because completely out of context for the final class user.

yes, for these classes it make sense.

Reply:
I pushed 2 last changes. One about const-ness, the other to hide some visit_ methods when they are not meant to be called as we discussed in this thread.